src/primitives/block.h

```diff
@@ -99,24 +103,54 @@ class CBlockHeader
 
     template <typename Stream, typename Operation>
     inline void SerializationOp(Stream& s, Operation ser_action) {
+        bool fCheckVeilDataHash = false;
         READWRITE(nVersion);
         READWRITE(hashPrevBlock);
-        READWRITE(hashVeilData);
+        if (nVersion >> 28 != 3) {
+            fCheckVeilDataHash = true;
+            // This is the version that the new PoW started at. This is when we stop using hashVeilData.
+            // We would normally use the nTime of the block. However, that is serialized after the hashVeilData
+            // So we don't have access to it
+            READWRITE(hashVeilData);
+        }
+
         READWRITE(nTime);
         READWRITE(nBits);
+
+        if (fCheckVeilDataHash && nTime >= nPowTimeStampActive) {
+            // We want to fail to accept this block. setting it NUll should fail all checks done on it
+            SetNull();
+            return;
+        }
+
+        if (nTime >= nPowTimeStampActive) {
+            READWRITE(hashMerkleRoot);
+            READWRITE(hashWitnessMerkleRoot);
+            READWRITE(hashAccumulators);
+        }
+
         if (nVersion & PROGPOW_BLOCK && nTime >= nPowTimeStampActive) {
             READWRITE(nHeight);
             READWRITE(nNonce64);
             READWRITE(mixHash);
         } else if (nVersion & RANDOMX_BLOCK && nTime >= nPowTimeStampActive) {
             READWRITE(nHeight);
             READWRITE(nNonce);
+        } else if (nVersion & SHA256D_BLOCK && nTime >= nPowTimeStampActive) {
+            READWRITE(nNonce64);
         } else {
             READWRITE(nNonce);
         }
+
```

Comment:
Rather than all these extra checks, this might be better all contained inside an nTime check:
```
if (nTime >= nPowTimeStampActive) {
  if (fCheckVeilDataHash) {
    ... return...
    }
  read merkle root
  read witness
  read accumulators
}
```
Since it's going to be one of the 3, I would think a switch would be better served.  Since you check the bit, what happens if you have a block that set more than one of the bits?  Should you not mask off the the type and switch it?
```
nPowType = (nVersion & (PROGPOW_BLOCK | RANDOMX_BLOCK | SHA256D_BLOCK))
switch (nPowType)
  case PROGPOW_BLOCK
...
  case 0
     READWRITE (nNonce)
  default: 
    // someone's setting more than one!
```
```

Reply:
Great catch. I am going to add code into the CheckBlockHeader that will enforce only one bit to be present in each block header.

Reply:
Updated the formating. Also added an additional check that wont allow blocks to have multiple algo bits set.